node_groups.tf

```diff
@@ -1,112 +1,29 @@
+# before other resources are ready. Removes race conditions
+data "null_data_source" "node_groups" {
+  count = var.create_eks ? 1 : 0
+
+  inputs = {
+    cluster_name = var.cluster_name
+
+    # Ensure these resources are created before "unlocking" the data source.
+    # `depends_on` causes a refresh on every run so is useless here.
+    # [Re]creating or removing these resources will trigger recreation of Node Group resources
+    aws_auth         = coalescelist(kubernetes_config_map.aws_auth[*].id, [""])[0]
+    role_NodePolicy  = coalescelist(aws_iam_role_policy_attachment.workers_AmazonEKSWorkerNodePolicy[*].id, [""])[0]
+    role_CNI_Policy  = coalescelist(aws_iam_role_policy_attachment.workers_AmazonEKS_CNI_Policy[*].id, [""])[0]
+    role_Container   = coalescelist(aws_iam_role_policy_attachment.workers_AmazonEC2ContainerRegistryReadOnly[*].id, [""])[0]
+    role_autoscaling = coalescelist(aws_iam_role_policy_attachment.workers_autoscaling[*].id, [""])[0]
   }
 }
 
-resource "aws_eks_node_group" "workers" {
-  for_each = var.create_eks ? local.node_groups : {}
-
-  node_group_name = join("-", [var.cluster_name, each.key, random_pet.node_groups[each.key].id])
-
-  cluster_name  = var.cluster_name
-  node_role_arn = lookup(each.value, "iam_role_arn", aws_iam_role.node_groups[0].arn)
-  subnet_ids    = lookup(each.value, "subnets", local.workers_group_defaults["subnets"])
-
-  scaling_config {
-    desired_size = lookup(each.value, "node_group_desired_capacity", local.workers_group_defaults["asg_desired_capacity"])
-    max_size     = lookup(each.value, "node_group_max_capacity", local.workers_group_defaults["asg_max_size"])
-    min_size     = lookup(each.value, "node_group_min_capacity", local.workers_group_defaults["asg_min_size"])
-  }
-
-  ami_type        = lookup(each.value, "ami_type", null)
-  disk_size       = lookup(each.value, "root_volume_size", null)
-  instance_types  = [lookup(each.value, "instance_type", null)]
-  labels          = lookup(each.value, "node_group_k8s_labels", null)
-  release_version = lookup(each.value, "ami_release_version", null)
-
-  dynamic "remote_access" {
-    for_each = [
-      for node_group in [each.value] : {
-        ec2_ssh_key               = node_group["key_name"]
-        source_security_group_ids = lookup(node_group, "source_security_group_ids", [])
-      }
-      if lookup(node_group, "key_name", "") != ""
-    ]
-
-    content {
-      ec2_ssh_key               = remote_access.value["ec2_ssh_key"]
-      source_security_group_ids = remote_access.value["source_security_group_ids"]
-    }
-  }
-
-  version = aws_eks_cluster.this[0].version
-
-  tags = lookup(each.value, "node_group_additional_tags", null)
-
-  lifecycle {
-    create_before_destroy = true
-  }
+module "node_groups" {
+  source                 = "./modules/node_groups"
+  create_eks             = var.create_eks
+  cluster_name           = coalescelist(data.null_data_source.node_groups[*].outputs["cluster_name"], [""])[0]
```

Comment:
Can't this just be `var.cluster_name`?

Reply:
This sets up a dependency on the `kubernetes_config_map.aws_auth` resource to avoid a race condition.